src/frontend/src/binder/set_expr.rs

```diff
@@ -194,6 +218,85 @@ impl Binder {
+    fn corresponding(
+        &self,
+        left: &BoundSetExpr,
+        right: &BoundSetExpr,
+        corresponding: Corresponding,
+    ) -> Result<(ColIndexMapping, ColIndexMapping)> {
+        let check_duplicate_name = |set_expr: &BoundSetExpr| {
+            let mut name2idx = HashMap::new();
+            for (idx, field) in set_expr.schema().fields.iter().enumerate() {
+                if name2idx.insert(field.name.clone(), idx).is_some() {
+                    return Err(ErrorCode::InvalidInputSyntax(format!(
+                        "duplicated column name `{}` in a select list of a set operation",
+                        field.name,
+                    )));
+                }
+            }
+            Ok(name2idx)
+        };
+
+        // Within the columns of both side, the same <column name> shall not
+        // be specified more than once.
+        let name2idx_l = check_duplicate_name(left)?;
+        let name2idx_r = check_duplicate_name(right)?;
+
+        let mut corresponding_col_idx_l = vec![];
+        let mut corresponding_col_idx_r = vec![];
+
+        if let Some(column_list) = corresponding.column_list() {
+            // The select list of the corresponding set operation should be in the order of <corresponding column list>
+            for column in column_list {
+                let col_name = column.real_value();
+                if let Some(idx_l) = name2idx_l.get(&col_name)
+                    && let Some(idx_r) = name2idx_r.get(&col_name)
+                {
+                    corresponding_col_idx_l.push(*idx_l);
+                    corresponding_col_idx_r.push(*idx_r);
+                } else {
+                    return Err(ErrorCode::InvalidInputSyntax(format!(
+                        "Every <column name> in the <corresponding column list> shall be a \
+                        <column name> of both left and right side. Missing column: `{}`",
+                        col_name
+                    ))
+                    .into());
+                }
+            }
+        } else {
+            // The select list of the corresponding set operation should be
+            // in the order that appears in the <column name>s of the left side.
+            for field in &left.schema().fields {
+                let col_name = &field.name;
+                if col_name != UNNAMED_COLUMN
+                    && let Some(idx_l) = name2idx_l.get(col_name)
+                    && let Some(idx_r) = name2idx_r.get(col_name)
+                {
+                    corresponding_col_idx_l.push(*idx_l);
+                    corresponding_col_idx_r.push(*idx_r);
+                }
+            }
+
+            if corresponding_col_idx_l.is_empty() {
+                return Err(ErrorCode::InvalidInputSyntax(
+                    "When CORRESPONDING is specified, at least one column of the left side shall have a <column name> that is the \
+                    <column name> of some column of the right side"
```

Comment:
Do you mean 
```suggestion
                    "When CORRESPONDING is specified, left and right side should have at least one common column name"
```
?

Reply:
> Why using <>? 

I can remove this.

Reply:
> Left or right side does not have column name, the columns in left or right have column names.

Sorry, I do not understand. Can you explain to me like I'm a normal user?

Reply:
Left or right side are `query`s. A `query` does not have column name. The columns/select items in a query have column name. 

> left and right side should have at least one common column name

This can be rephrased to `left and right side should have at least one column with identical column names`

The original one I use is from the  SQL 92 doc

Reply:
> A query does not have column name. The columns/select items in a query have column name.

Well, this sounds too language lawyer to me.. 🤡
I think "A query's column name" is clear enough to me. Do you really think its vague? 🤔 (Specification needs to be more accurate though)

Reply:
> Do you really think its vague?

I can definitely understand that expression perfectly. But using more accurate language can avoid ambiguity for users, especially CORRESPONDING is not a wildly available feature is other DB.